src/en/mangamo/src/eu/kanade/tachiyomi/extension/en/mangamo/Mangamo.kt

```diff
@@ -0,0 +1,435 @@
+                    ChapterDto::seriesId.name,
+                    ChapterDto::chapterNumber.name,
+                    ChapterDto::name.name,
+                    ChapterDto::createdAt.name,
+                    ChapterDto::onlyTransactional.name,
+                )
+
+                orderBy = listOf(descending(ChapterDto::chapterNumber.name))
+            },
+        ).asObservableSuccess().map { response ->
+            response.body.string().parseJson<QueryResultDto<ChapterDto>>().elements
+        }
+
+        val hideCoinChapters = coinMangaPref.contains(MangamoConstants.HIDE_COIN_MANGA_OPTION_CHAPTERS)
+
+        return Observable.combineLatest(seriesObservable, chaptersObservable) { series, chapters ->
+            chapters
+                .mapNotNull { chapter ->
+                    if (chapter.enabled != true) {
+                        return@mapNotNull null
+                    }
+
+                    val isFreeChapter = chapter.chapterNumber!! <= (series.maxFreeChapterNumber ?: 0)
+
+                    val isCoinChapter = chapter.onlyTransactional == true ||
+                        (series.onlyTransactional == true && !isFreeChapter)
+
+                    if (hideCoinChapters && isCoinChapter) {
+                        return@mapNotNull null
+                    }
+
+                    SChapter.create().apply {
+                        chapter_number = chapter.chapterNumber
+                        date_upload = chapter.createdAt!!
+                        name = chapter.name +
+                            if (isFreeChapter) {
+                                " (Free)"
+                            } else if (isCoinChapter) {
+                                " (Coin)"
+                            } else {
+                                ""
+                            }
+                        url = helper.getChapterUrl(chapter)
+                    }
+                }
+        }
+    }
+
+    override fun chapterListParse(response: Response): List<SChapter> =
+        throw UnsupportedOperationException()
+
+    private fun getPagesImagesRequest(series: Int, chapter: Int): Request {
+        return POST(
+            "${MangamoConstants.FIREBASE_FUNCTION_BASE_PATH}/page/$series/$chapter",
+            helper.jsonHeaders,
+            "{\"idToken\":\"${auth.getIdToken()}\"}".toRequestBody(),
+        )
+    }
+
+    override fun pageListRequest(chapter: SChapter): Request {
+        val uri = chapter.url.toHttpUrl()
+
+        val seriesId = uri.queryParameter(MangamoConstants.SERIES_QUERY_PARAM)!!.toInt()
+        val chapterId = uri.queryParameter(MangamoConstants.CHAPTER_QUERY_PARAM)!!.toInt()
+
+        return getPagesImagesRequest(seriesId, chapterId)
+    }
+
+    override fun pageListParse(response: Response): List<Page> {
+        val data = response.body.string().parseJson<List<PageDto>>()
+
+        return data.map {
+            Page(it.pageNumber - 1, imageUrl = it.uri)
+        }.sortedBy { it.index }
+    }
```

Comment:
should throw custom exception when user tries reading paid chapters without payment; maybe in an interceptor detect http 401?

Reply:
Added phrasing for this. It's a little more vague than talking about payment since there are a number of reasons someone might not be able to access a chapter, but it is more comprehensible than seeing a 401.